src/wallet/walletdb.cpp

```diff
@@ -1210,36 +1213,50 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
-    } catch (...) {
-        result = DBErrors::CORRUPT;
-    }
+        // Upgrade all of the descriptor caches to cache the last hardened xpub
+        // This operation is not atomic, but if it fails, only new entries are added so it is backwards compatible
+        try {
+            pwallet->UpgradeDescriptorCache();
+        } catch (...) {
+            result = DBErrors::CORRUPT;
+        }
 
-    // Since it was accidentally possible to "encrypt" a wallet with private keys disabled, we should check if this is
-    // such a wallet and remove the encryption key records to avoid any future issues.
-    // Although wallets without private keys should not have *ckey records, we should double check that.
-    // Removing the mkey records is only safe if there are no *ckey records.
-    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && pwallet->HasEncryptionKeys() && !pwallet->HaveCryptedKeys()) {
-        pwallet->WalletLogPrintf("Detected extraneous encryption keys in this wallet without private keys. Removing extraneous encryption keys.\n");
-        for (const auto& [id, _] : pwallet->mapMasterKeys) {
-            if (!EraseMasterKey(id)) {
-                pwallet->WalletLogPrintf("Error: Unable to remove extraneous encryption key '%u'. Wallet corrupt.\n", id);
-                return DBErrors::CORRUPT;
+        // Since it was accidentally possible to "encrypt" a wallet with private keys disabled, we should check if this is
+        // such a wallet and remove the encryption key records to avoid any future issues.
+        // Although wallets without private keys should not have *ckey records, we should double check that.
+        // Removing the mkey records is only safe if there are no *ckey records.
+        if (pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && pwallet->HasEncryptionKeys() && !pwallet->HaveCryptedKeys()) {
+            pwallet->WalletLogPrintf("Detected extraneous encryption keys in this wallet without private keys. Removing extraneous encryption keys.\n");
+            for (const auto& [id, _] : pwallet->mapMasterKeys) {
+                if (!EraseMasterKey(id)) {
+                    pwallet->WalletLogPrintf("Error: Unable to remove extraneous encryption key '%u'. Wallet corrupt.\n", id);
+                    return DBErrors::CORRUPT;
+                }
             }
+            pwallet->mapMasterKeys.clear();
+        }
+    } else {
+        // In read-only mode, log what we would have done
+        if (!upgraded_txs.empty()) {
+            pwallet->WalletLogPrintf("Skipping write of %d upgraded transactions in read-only mode\n", upgraded_txs.size());
+        }
+        if (!has_last_client || last_client != CLIENT_VERSION) {
+            pwallet->WalletLogPrintf("Skipping client version update in read-only mode\n");
+        }
+        if (any_unordered) {
+            pwallet->WalletLogPrintf("Skipping transaction reordering in read-only mode\n");
```

Comment:
In f6f1f4199b4a3c2d7ef7c1bcfa122a9d8c5eeb64 "wallet: Allow read-only database access for info and dump commands"

I don't think we need to log for these, especially since we don't log that they happen in the first place.